### ISSUE TITLE
Is Borg really implemented right ?

### ISSUE THREAD

**Issue by jpic** [faif/python-patterns#3]
I thought that the point of Borg was to allow subclassing. However, the implementation doesn't match expectations in this case:

```
class Borg:
    __shared_state = {}

    def __init__(self):
        self.__dict__ = self.__shared_state
        self.state = 'Running'

    def __str__(self):
        return self.state

instance0 = Borg()
instance1 = Borg()

instance0.state = 'Idle'

print instance1  # prints 'Idle', instead of 'Running', as expected

borg = Borg()
borg.state = 'Idle'


class YourBorg(Borg):
    pass

borg = YourBorg()
print borg  # prints 'Running' instead of 'Idle', **not** as expected
```

Are you sure that Borg supports setting attributes in the constructor ?

It looks like it defeats the purpose of Borg...

**Comment by modasserbillah** [comment]
I know it's closed but the borg pattern still sets the state during init and the issue pointed out by @jpic stands. [Reference](https://github.com/faif/python-patterns/blob/master/patterns/creational/borg.py#L40)

Am I missing something here?

**Comment by faif** [comment]
Hi. 

It does set a state during `init` but that is not required. The state is changed later anyway, and subclassing works as expected (the state is modified also for subclasses). What is your concern?

**Comment by modasserbillah** [comment]
My concern is when an instance of the subclass is instantiated, the state gets changed. In my opinion, that should not be the case. When a new instance is created, the state should not change unless explicitly changed. Here's the argument from Alex Martelli : https://learning.oreilly.com/library/view/python-cookbook/0596001673/ch05s23.html

**Comment by faif** [comment]
I see your point. Setting the state in `__init__` was used as an example but indeed it might not give the right idea. Feel free to make the change and I'll merge it.

**Comment by faif** [comment]
Fixed in #338
